### PR TITLE
add support for building swagger/OpenAPI JSON

### PR DESCRIPTION
**Pull request by ryanpetrello**
to build, run `make swagger`; a file named `swagger.json` will be
written to the current working directory

```shell
$ make swagger
$ SWAGGER_RELEASE_VERSION=x.y.z make swagger
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -365,6 +365,12 @@ pyflakes: reports
 pylint: reports
 	@(set -o pipefail && $@ | reports/$@.report)
 
+swagger: reports
+	@if [ "$(VENV_BASE)" ]; then \
+		. $(VENV_BASE)/awx/bin/activate; \
+	fi; \
+	(set -o pipefail && py.test awx/main/tests/docs --release=$(RELEASE_VERSION) | tee reports/$@.report)
```
Comment (ryanpetrello) [inline Makefile]: The docs are built via a functional Python test that:

1.  Does an `HTTP GET` to a new `settings.DEBUG`-only Swagger view (powered by the `django-rest-swagger` library) via the same virtual WSGI `get()` fixture that other functional tests use.
2. Does various transformations with the response body.
3. Writes the results JSON file, `swagger.json`.

File: awx/api/swagger.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        try:
+            serializer = self.view.get_serializer()
+        except Exception:
+            serializer = None
+            warnings.warn('{}.get_serializer() raised an exception during '
+                          'schema generation. Serializer fields will not be '
+                          'generated for {} {}.'
+                          .format(self.view.__class__.__name__, method, path))
+
+        # auto-generate a topic/tag for the serializer based on its model
```
Comment (ryanpetrello) [inline awx/api/swagger.py]: Look at the model or the view to determine a "category" for the current method and path.  This is usually something like "Job Templates".

File: awx/api/templates/api/base_variable_data.md [hunk 3]
```diff
@@ -1,9 +1,13 @@
+{% ifmeth GET %}
```
Comment (ryanpetrello) [inline awx/api/templates/api/base_variable_data.md:1]: This is a new custom template tag we can use to compose Swagger docs from our existing API browser Markdown.  The idea is that as `django-rest-swagger` builds the API descriptions from this markdown, it does so for _each_ potential request method.  So `django-rest-swagger` will render this description template (potentially) five times: `GET`, `POST`, `PUT`, `PATCH`, and `DELETE`.

screen shot 2018-02-02 at 12 21 49 pm

The `ifmeth` tag allows you to provide hints to the template rendering to filter out blocks that aren't applicable to the current method it's generating content for.

I've added this to most of the _base_ markdown templates, but there's lots of other places where we've overridden templates that I still need to add these.

File: awx/api/templates/api/group_all_hosts_list.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# List All {{ model_verbose_name_plural|title }} for this {{ parent_model_verbose_name|title }}:
```
Comment (ryanpetrello) [inline awx/api/templates/api/group_all_hosts_list.md:1]: `anora` is a new filter you'll see sprinkled everywhere that adds "for a/an X" before the variable; it looked nicer in the generated docs than "for this X"

File: awx/api/templates/swagger/config.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+---
```
Comment (ryanpetrello) [inline awx/api/templates/swagger/config.yml:1]: This `.yml` file gives the docs team the ability to control the ordering of the categories in the generated Swagger doc.  We may just use this to tweak the output the first few times.

File: awx/api/templates/swagger/description.md [hunk 6]
```diff
@@ -0,0 +1 @@
+The Ansible Tower API Reference Manual provides in-depth documentation for Tower's REST API, including examples on how to integrate with it.
```
Comment (ryanpetrello) [inline awx/api/templates/swagger/description.md:1]: This is markdown the docs team can customize to provide a sort of intro paragraph or two at the top of the docs.

File: awx/main/tests/docs/conftest.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+from awx.main.tests.functional.conftest import *  # noqa
+
+
+def pytest_addoption(parser):
```
Comment (ryanpetrello) [inline awx/main/tests/docs/conftest.py:4]: This allows you to invoke the doc build test with `py.test awx/main/tests/docs --release=3.3.0`.  It falls back to the awx git version tag.

**Comment by ryanpetrello** [comment]
TODO:

- Need to go through all the API `.md` templates and sprinkle `{% ifmeth %}` where appropriate.
- OpenAPI supports marking deprecated paths; add smarts to generate that based on our `APIView.deprecated` attribute.
- One notable that's missing is payload examples (e.g., example JSON content for `HTTP POST/PUT/PATCH`); it might be nice to expand on the tests here to preload a bunch of data and write https://swagger.io/specification/#requestBodyObject objects

File: awx/api/templates/swagger/config.yml (hunk 5 above)
Comment (ryanpetrello) [reply awx/api/templates/swagger/config.yml:1]: @tvo318 @docschick ^

File: awx/api/templates/swagger/description.md (hunk 6 above)
Comment (ryanpetrello) [reply awx/api/templates/swagger/description.md:1]: @tvo318 @docschick ^

File: shippable.yml [hunk 8]
```diff
@@ -7,6 +7,7 @@ env:
   - AWX_BUILD_TARGET=test
   - AWX_BUILD_TARGET=ui-test-ci
   - AWX_BUILD_TARGET="flake8 jshint"
+  - AWX_BUILD_TARGET="swagger"
```
Comment (matburt) [inline shippable.yml:10]: Do we want to run this on every shippable run or just on a nightly?

File: awx/main/templatetags/swagger.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import re
+from django.utils.encoding import force_unicode
+from django import template
+
+register = template.Library()
+
+CONSONANT_SOUND = re.compile(r'''one(![ir])''', re.IGNORECASE|re.VERBOSE)  # noqa
```
Comment (matburt) [inline awx/main/templatetags/swagger.py:7]: ❤️

**Review by matburt — APPROVED** [review]

**Pull request merged**